src/main/java/org/jenkinsci/plugins/KeycloakAccess.java

```diff
@@ -103,6 +102,13 @@ public UserDetails authenticate( String username, String password ) throws Authe
 		}
 	}
 
+	/**
+	 * Get the roles from the user, from either cache or calling Keycloak.
+	 * @param username - username to retrieve keycloak roles
+	 * @param token - Null token will get client token for role retrieval
+	 * @return
+	 * @throws IOException
+	 */
 	private String[] getRolesForUser( String username, String token ) throws IOException {
 		if (cache.isEnabled()) {
 			Collection<String> cacheRoles = cache.getRolesForUser( username );

```

Comment:
From later comment, possible to add in functionality to add invalid users too? Maybe have it check for existence of username in cache here and then next step would be to get list of roles back -- if roles is empty/null then send back usernotfound / dataretrieval exception.

Reply:
I added caching here though for invalid users